Avoid fast path mask left-align check in compiled TransformerEncoder

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #163773

Fixes #163640

This PR avoids a mask left align check in the case that we're operating under torch.compile / torch.export. Originally, I planned to make a more invasive change to auto-disable the fast path entirely underneath torch.compile / torch.export, but I realized during testing that the fast path wasn't actually causing compile issues outside of the narrow issue identified here.

jbschlosser committed:
Avoid fast path mask left-align check in compiled TransformerEncoder

[ghstack-poisoned]

mikaylagawarecki reviewed (approved):
Thank you, just one question from me

mikaylagawarecki commented on torch/nn/modules/transformer.py:
```diff
@@ -462,6 +462,14 @@ def forward(
+        # This check avoids a call to torch._nested_tensor_from_mask_left_aligned() that
+        # breaks in torch.compile.
+        elif (
+            (not hasattr(self, "mask_check")) or self.mask_check
```
I know this is used in the branch just below, but to confirm, if not hasattr(self, "mask_check"), this means the mask is always checked?

jbschlosser replied:
yes that's correct.. the default is to do a mask check if for some reason the attribute isn't present. Doing a mask check is also the default if the attribute was unspecified upon construction.

jbschlosser committed:
Update on "Avoid fast path mask left-align check in compiled TransformerEncoder"


Fixes #163640

This PR avoids a mask left align check in the case that we're operating under torch.compile / torch.export. Originally, I planned to make a more invasive change to auto-disable the fast path entirely underneath torch.compile / torch.export, but I realized during testing that the fast path wasn't actually causing compile issues outside of the narrow issue identified here.

[ghstack-poisoned]
